apps/browser/bookmarks/bookmarkfiltermodel.cpp

```diff
@@ -28,24 +28,53 @@ bool BookmarkFilterModel::filterAcceptsRow(int sourceRow, const QModelIndex &sou
 {
     QModelIndex index = sourceModel()->index(sourceRow, 0, sourceParent);
 
-    if (sourceModel()->data(index, DeclarativeBookmarkModel::UrlRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)
-            || sourceModel()->data(index, DeclarativeBookmarkModel::TitleRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)) {
+    if ((sourceModel()->data(index, DeclarativeBookmarkModel::UrlRole).toString().trimmed().contains(search(), Qt::CaseInsensitive)
+         || sourceModel()->data(index, DeclarativeBookmarkModel::TitleRole).toString().trimmed().contains(search(), Qt::CaseInsensitive))
+         && m_countFilterAccepts < m_maxDisplayedItems) {
+        m_countFilterAccepts++;
         return true;
     }
     return false;
 }
 
+void BookmarkFilterModel::setSourceModel(QAbstractItemModel *sourceModel)
+{
+    if (sourceModel) {
+        m_countFilterAccepts = 0;
+        m_maxDisplayedItems = sourceModel->rowCount();
+        QSortFilterProxyModel::setSourceModel(sourceModel);
+        invalidate();
```

Comment:
According to [the docs](https://doc.qt.io/qt-5/qabstractproxymodel.html#setSourceModel):
> Subclasses should call beginResetModel() at the beginning of the method, disconnect from the old model, call this method, connect to the new model, and call endResetModel().

so presumably you should be doing this here rather than invalidating.